From a6cc144ffc71cf3ef060c1e1a16e4b8178e2fd52 Mon Sep 17 00:00:00 2001
From: Shammah Chancellor <shammah.chancellor@gmail.com>
Date: Tue, 25 May 2021 10:13:30 -0700
Subject: [PATCH 112/196] Implement a new simplified mining API
 `getunsolvedblock` (#185)

This API takes an address and returns a target, and the serialized block
as hex. Because the nonce was expanded, there is no need to grind on the
coinbase in order to produce a valid block. Thus, we can implement a
miner which takes only the block header. However, currently we also
include the transactions as there is no internal caching for the
transaction set. These are needed in submit block for validating the
merkle root and broadcasting the transactions later. Mining software
will need to hold on to them, but does not need to do anything with
them.
---
 src/rpc/mining.cpp                            | 84 +++++++++++++++++++
 .../logos_rpc_getrawunsolvedblock.py          | 61 ++++++++++++++
 2 files changed, 145 insertions(+)
 create mode 100755 test/functional/logos_rpc_getrawunsolvedblock.py

diff --git a/src/rpc/mining.cpp b/src/rpc/mining.cpp
index c982927c5..bb6f3aa9a 100644
--- a/src/rpc/mining.cpp
+++ b/src/rpc/mining.cpp
@@ -11,6 +11,7 @@
 #include <config.h>
 #include <consensus/activation.h>
 #include <consensus/consensus.h>
+#include <consensus/merkle.h>
 #include <consensus/params.h>
 #include <consensus/validation.h>
 #include <core_io.h>
@@ -1006,6 +1007,88 @@ static UniValue getblocktemplate(const Config &config,
     return result;
 }
 
+static UniValue getrawunsolvedblock(const Config &config,
+                                    const JSONRPCRequest &request) {
+    RPCHelpMan{
+        "getrawunsolvedblock",
+        "Get a serialized block candidate with coinbase sending to the "
+        "specified address",
+        {
+            {"address", RPCArg::Type::STR, RPCArg::Optional::NO,
+             "The address to send the newly generated coins to."},
+        },
+        RPCResult{
+            RPCResult::Type::STR_HEX,
+            "blockhex",
+            "Serialized block as hex.",
+        },
+        RPCExamples{"\nCreate an unsolved block to to myaddress\n" +
+                    HelpExampleCli("getrawunsolvedblock", "\"myaddress\"") +
+                    "If you are running the Lotus wallet, you can get a new "
+                    "address to send the newly generated coins to with:\n" +
+                    HelpExampleCli("getnewaddress", "")},
+    }
+        .Check(request);
+
+    CTxDestination destination =
+        DecodeDestination(request.params[0].get_str(), config.GetChainParams());
+    if (!IsValidDestination(destination)) {
+        throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY,
+                           "Error: Invalid address");
+    }
+
+    const CChainParams &chainparams = config.GetChainParams();
+
+    NodeContext &node = EnsureNodeContext(request.context);
+    if (!node.connman) {
+        throw JSONRPCError(
+            RPC_CLIENT_P2P_DISABLED,
+            "Error: Peer-to-peer functionality missing or disabled");
+    }
+
+    if (node.connman->GetNodeCount(CConnman::CONNECTIONS_ALL) == 0) {
+        throw JSONRPCError(RPC_CLIENT_NOT_CONNECTED, "Node is not connected!");
+    }
+
+    if (::ChainstateActive().IsInitialBlockDownload()) {
+        throw JSONRPCError(RPC_CLIENT_IN_INITIAL_DOWNLOAD, PACKAGE_NAME
+                           " is in initial sync and waiting for blocks...");
+    }
+
+    LOCK(cs_main);
+    const CTxMemPool &mempool = EnsureMemPool(request.context);
+
+    // Store the pindexBest used before CreateNewBlock, to avoid races
+    CBlockIndex *pindexPrevNew = ::ChainActive().Tip();
+
+    // Create new block with provided coinbase output script.
+    CScript coinbase_script = GetScriptForDestination(destination);
+    std::unique_ptr<CBlockTemplate> pblocktemplate =
+        BlockAssembler(config, mempool).CreateNewBlock(coinbase_script);
+    if (!pblocktemplate) {
+        throw JSONRPCError(RPC_OUT_OF_MEMORY, "Out of memory");
+    }
+
+    // Need to update only after we know CreateNewBlock succeeded
+    CHECK_NONFATAL(pindexPrevNew);
+
+    // pointer for convenience
+    CBlock *pblock = &pblocktemplate->block;
+    // Update nTime (can update nBits on testnet)
+    UpdateTime(pblock, chainparams, pindexPrevNew);
+    pblock->nNonce = 0;
+    // Update merkle root so mining software need not touch it.
+    pblock->hashMerkleRoot = BlockMerkleRoot(*pblock);
+
+    CDataStream ssBlock(SER_NETWORK, PROTOCOL_VERSION);
+    ssBlock << *pblock;
+    arith_uint256 hashTarget = arith_uint256().SetCompact(pblock->nBits);
+    UniValue result(UniValue::VOBJ);
+    result.pushKV("blockhex", HexStr(ssBlock));
+    result.pushKV("target", hashTarget.GetHex());
+    return result;
+}
+
 class submitblock_StateCatcher final : public CValidationInterface {
 public:
     uint256 hash;
@@ -1160,6 +1243,7 @@ void RegisterMiningRPCCommands(CRPCTable &t) {
         {"mining",     "getmininginfo",         getmininginfo,         {}},
         {"mining",     "prioritisetransaction", prioritisetransaction, {"txid", "dummy", "fee_delta"}},
         {"mining",     "getblocktemplate",      getblocktemplate,      {"template_request"}},
+        {"mining",     "getrawunsolvedblock",   getrawunsolvedblock,   {"address"}},
         {"mining",     "submitblock",           submitblock,           {"hexdata", "dummy"}},
         {"mining",     "submitheader",          submitheader,          {"hexdata"}},
 
diff --git a/test/functional/logos_rpc_getrawunsolvedblock.py b/test/functional/logos_rpc_getrawunsolvedblock.py
new file mode 100755
index 000000000..8ddc8eadd
--- /dev/null
+++ b/test/functional/logos_rpc_getrawunsolvedblock.py
@@ -0,0 +1,61 @@
+#!/usr/bin/env python3
+# Copyright (c) 2021 The Logos Foundation
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""
+Tests for Lotus `getrawunsolvedblock` rpc call
+"""
+
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.util import (
+    assert_equal,
+    assert_array_result,
+)
+from test_framework.messages import (
+    CBlock,
+    FromHex,
+)
+
+from decimal import Decimal
+
+NUM_TRANSACTIONS = 10
+
+
+class LotusGetRawUnsolvedBlock(BitcoinTestFramework):
+    def set_test_params(self):
+        self.setup_clean_chain = True
+        self.num_nodes = 2
+
+    def run_test(self):
+        node = self.nodes[0]
+        address = node.get_deterministic_priv_key().address
+        # Generate generate some blocks and mature the coinbases
+        node.generatetoaddress(101, address)
+        last_generated_block = node.getbestblockhash()
+        txids = [node.sendtoaddress(node.getnewaddress(), 1)
+                 for x in range(NUM_TRANSACTIONS)]
+        response = node.getrawunsolvedblock(address)
+        block = FromHex(CBlock(), response['blockhex'])
+        # coinbase and the 3 transactions we sent
+        assert_equal(block.hashMerkleRoot, block.calc_merkle_root())
+        # Should have all the transactions, plus the coinbase tx
+        assert_equal(len(block.vtx), NUM_TRANSACTIONS+1)
+        # Skip coinbase. Transactions should be sorted by Id
+        txids.sort()
+        for idx in range(NUM_TRANSACTIONS):
+            # All transactions found and in order
+            assert_equal(txids[idx], block.vtx[idx+1].txid_hex)
+        block.solve()
+        node.submitblock(block.serialize().hex())
+        response = node.getbestblockhash()
+        assert_equal(response, block.hash)
+        # Ensure the transactions are confirmed and in the wallet
+        txns = node.listsinceblock(last_generated_block)['transactions']
+        for idx in range(NUM_TRANSACTIONS):
+            assert_array_result(txns,
+                                {'txid': txids[idx]},
+                                {'blockhash': block.hash, 'confirmations': 1})
+
+
+if __name__ == '__main__':
+    LotusGetRawUnsolvedBlock().main()
-- 
2.39.5 (Apple Git-154)

